### PR TITLE
perf/security: scan engine hardening — all 7 goals from multi-model code review (#103)

### LINKED ISSUE #102: perf/security: scan engine hardening from multi-model code review
## Summary

Port 7 targeted improvements from a structured multi-model code review (5 AI models, 22 evaluation goals, Karpathy autoresearch pattern) into the main script. These changes were validated in isolation with full audit logging and before/after timing benchmarks.

## Performance: 15.6% Faster Scans

**How we measured it:** Baseline and final timing used 3 consecutive runs of a single-region JsonOutput scan (`-Region eastus -NoPrompt -JsonOutput -SkipRegionValidation`) averaged to reduce variance. Baseline average: 15,895ms. Final average: 13,411ms. Improvement: -2,484ms (-15.6%).

**What changed:** The scan engine previously fetched SKU data and quota data sequentially — one REST call finished before the next started. The improvement uses .NET HttpClient to fire both first-page requests concurrently (`Task.WaitAll`), then pages sequentially after. Since SKU and quota APIs are independent and the first page is the largest payload, overlapping their network latency cuts wall-clock time by ~2.5 seconds per region.

## Security: ReDoS Elimination

User-supplied SKU filter patterns (`-SkuFilter`) were previously compiled as .NET regex. Malicious or accidental patterns like `(a+)+$` could cause catastrophic backtracking. Replaced with PowerShell `-like` operator which has zero ReDoS surface, handles `*` and `?` wildcards natively, and is faster for simple pattern matching. Added 128-char limit and character whitelist validation.

### Performance
- **HttpClient concurrent first-page fetch** for SKU+quota (15.6% faster) — Source: Claude Sonnet
- **Get-CapValue PSObject.Properties defensive check** — handles edge-case '0'/'False' capability values — Source: Claude Sonnet

### Security
- **-like operator replaces regex** in Test-SkuMatchesFilter + inline scan blocks — Source: GPT-5.4

### Resilience & Correctness
- **HTTP 500 (InternalServerError)** added to Invoke-WithRetry retry pattern — Source: GPT-5.4
- **Per-subscription scan timer** excludes one-time pricing phase for accurate timing — Source: GPT-5.4
- **Write-Host gating** via module-qualified override with runtime $SuppressConsole — Source: Claude Sonnet
- **Pipeline emit dual-flag gate** prevents double output in -JsonOutput mode — Source: GPT-5.4

## Methodology

5 AI models (Claude Opus 4.6, Claude Sonnet 4.6, GPT-5.3 Codex, GPT-5.4, GPT-5.4 Mini) each independently implemented the full script. 22 improvement goals were evaluated across 3 tiers:
- **Tier 1 (timing-decisive):** Apply both versions, keep the faster one
- **Tier 2 (binary PASS/FAIL):** Verify already present in baseline
- **Tier 3 (human judgment):** Apply the recommended model's implementation, test

15 of 22 goals were already present in the baseline. 7 were applied from 2 winning models (Sonnet: 3 wins, GPT-5.4: 4 wins). 0 reverted. Full audit trail in scratch/Tournament/.

### LINKED ISSUE #103: perf/security: scan engine hardening from multi-model code review
## Summary

Port 7 targeted improvements from a structured multi-model code review (5 AI models, 22 evaluation goals, Karpathy autoresearch pattern) into the main script. These changes were validated in isolation with full audit logging and before/after timing benchmarks.

## Performance: 15.6% Faster Scans

**How we measured it:** Baseline and final timing used 3 consecutive runs of a single-region JsonOutput scan (`-Region eastus -NoPrompt -JsonOutput -SkipRegionValidation`) averaged to reduce variance. Baseline average: 15,895ms. Final average: 13,411ms. Improvement: -2,484ms (-15.6%).

**What changed:** The scan engine previously fetched SKU data and quota data sequentially — one REST call finished before the next started. The improvement uses .NET HttpClient to fire both first-page requests concurrently (`Task.WaitAll`), then pages sequentially after. Since SKU and quota APIs are independent and the first page is the largest payload, overlapping their network latency cuts wall-clock time by ~2.5 seconds per region.

## Security: ReDoS Elimination

User-supplied SKU filter patterns (`-SkuFilter`) were previously compiled as .NET regex. Malicious or accidental patterns like `(a+)+$` could cause catastrophic backtracking. Replaced with PowerShell `-like` operator which has zero ReDoS surface, handles `*` and `?` wildcards natively, and is faster for simple pattern matching. Added 128-char limit and character whitelist validation.

### Performance
- **HttpClient concurrent first-page fetch** for SKU+quota (15.6% faster) — Source: Claude Sonnet
- **Get-CapValue PSObject.Properties defensive check** — handles edge-case '0'/'False' capability values — Source: Claude Sonnet

### Security
- **-like operator replaces regex** in Test-SkuMatchesFilter + inline scan blocks — Source: GPT-5.4

### Resilience & Correctness
- **HTTP 500 (InternalServerError)** added to Invoke-WithRetry retry pattern — Source: GPT-5.4
- **Per-subscription scan timer** excludes one-time pricing phase for accurate timing — Source: GPT-5.4
- **Write-Host gating** via module-qualified override with runtime $SuppressConsole — Source: Claude Sonnet
- **Pipeline emit dual-flag gate** prevents double output in -JsonOutput mode — Source: GPT-5.4

## Methodology

5 AI models (Claude Opus 4.6, Claude Sonnet 4.6, GPT-5.3 Codex, GPT-5.4, GPT-5.4 Mini) each independently implemented the full script. 22 improvement goals were evaluated across 3 tiers:
- **Tier 1 (timing-decisive):** Apply both versions, keep the faster one
- **Tier 2 (binary PASS/FAIL):** Verify already present in baseline
- **Tier 3 (human judgment):** Apply the recommended model's implementation, test

15 of 22 goals were already present in the baseline. 7 were applied from 2 winning models (Sonnet: 3 wins, GPT-5.4: 4 wins). 0 reverted. Full audit trail in scratch/Tournament/.

### PR DESCRIPTION
**Pull request by ZacharyLuz**
## Description

Ports all 7 targeted improvements from a structured multi-model code review (5 AI models, 22 evaluation goals). See [docs/MULTI-MODEL-CODE-REVIEW.md](docs/MULTI-MODEL-CODE-REVIEW.md) for full methodology.

Closes #102
Closes #103

### Performance
- **HttpClient concurrent first-page fetch** (Goal 6 ΓÇö Claude Sonnet): Replaced Az cmdlet scan engine with direct REST API calls. HttpClient fires SKU and quota first-page requests concurrently via `Task.WaitAll`, then pages remaining results sequentially. Adds `ConvertFrom-RestSku`/`ConvertFrom-RestQuota` normalization helpers. Bearer token via `Get-AzAccessToken`, zeroed after use. **Live tested: 9.2s -> 7.3s (20.6% faster).**
- **Get-CapValue defensive check** (Goal 7 ΓÇö Claude Sonnet): `PSObject.Properties` check with `_CapIndex` O(1) fast path replaces truthy check that failed for '0'/'False' cap values.

### Security
- **ReDoS elimination** (Goal 14 ΓÇö GPT-5.4): Replaced regex-based SKU filter matching with `-like` operator in `Test-SkuMatchesFilter` and both inline scan blocks. Added 128-char length limit and character whitelist validation. Zero ReDoS attack surface.

### Resilience
- **HTTP 500 retry** (Goal 3 ΓÇö GPT-5.4): Added `InternalServerError` to `Invoke-WithRetry` pattern. Catches .NET enum name in addition to status code and message text.
- **Per-subscription scan timer** (Goal 12 ΓÇö GPT-5.4): Added `$subscriptionScanStartTime` per subscription. Scan elapsed time no longer includes one-time pricing phase.

### Output
- **Write-Host gating** (Goal 16 ΓÇö Claude Sonnet): Module-qualified `Write-Host` override with `$script:SuppressConsole` runtime flag. All 349 Write-Host calls become no-ops when `-JsonOutput` is active. Reversible at runtime.
- **Pipeline emit gate** (Goal 17 ΓÇö GPT-5.4): `$familyDetails` emitted to pipeline only when `[Console]::IsOutputRedirected` is true. Blocked in `-JsonOutput` mode to prevent double output.

### Documentation
- **Multi-model code review write-up**: Methodology, performance measurement, model scorecard.

## Validation

- [x] Syntax validation after each commit
- [x] Pester: 178 passed, 0 failed
- [x] PSScriptAnalyzer: 2 advisory warnings (expected from Write-Host override)
- [x] Live Azure scan: 9.2s -> 7.3s with REST+HttpClient engine
- [x] Live JsonOutput: clean JSON, zero Write-Host leakage

## Quality Checklist

- [x] **No AI instructional comments**
- [x] **No empty catch blocks**
- [x] **Version strings in sync** (no version bump ΓÇö deferred to release)
- [x] **PSScriptAnalyzer clean** (2 advisory warnings, both expected)
- [x] **Pester tests pass**
- [x] **Syntax valid**

### CODE REVIEW

File: Get-AzVMAvailability.ps1 [hunk 1]
```diff
@@ -352,6 +352,22 @@ param(
...
+# Console suppression: always-defined Write-Host override with runtime flag check.
+# Module-qualified delegation preserves original behavior when not suppressed.
+$script:SuppressConsole = $false
+function Write-Host {
+    param(
+        [Parameter(Position = 0, ValueFromPipeline)]
+        [object]$Object = '',
+        [System.ConsoleColor]$ForegroundColor,
+        [System.ConsoleColor]$BackgroundColor,
+        [switch]$NoNewline
```
[the remark this answers is not in the document]
Comment (ZacharyLuz) [reply Get-AzVMAvailability.ps1]: **Partially agree — deferred to v2.0.**

The proxy is incomplete: it omits `-Separator` and `ValueFromRemainingArguments`. However, auditing all ~700 `Write-Host` calls in this script confirms none pass multi-object arrays or `-Separator`. In practice the incomplete signature causes zero behavioral difference today.

The v2.0 module conversion will replace this with a full proxy (including `[CmdletBinding()]`, `ValueFromRemainingArguments=$true`, and proper delegation to `Microsoft.PowerShell.Utility\Write-Host`). Tracked in the v2.0 roadmap.

File: Get-AzVMAvailability.ps1 [hunk 2]
```diff
@@ -3693,6 +3723,12 @@ if ($JsonOutput) {
     return
 }
 
+# Emit structured objects to pipeline only when output is redirected (piped, assigned to variable, or sent to file).
+# In interactive terminal mode, objects are suppressed to preserve the clean Write-Host UX.
```
[the remark this answers is not in the document]
Comment (ZacharyLuz) [reply Get-AzVMAvailability.ps1]: **Agree — comment was misleading. Fixed in this PR.**

The comment previously claimed the gate fires when output is "piped, assigned to variable, or sent to file" — which is incorrect. `[Console]::IsOutputRedirected` only detects console-level stdout redirection (e.g., `> file.txt`), not PowerShell pipeline usage or variable assignment.

Comment updated to: *"detects console-level redirection only — does NOT detect PS pipeline usage. For interactive pipeline scenarios, use `-JsonOutput`. A dedicated `-PassThru` switch is planned for v2.0."*

File: tests/WriteHost-Gating.Tests.ps1 [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+BeforeAll {
+    Import-Module "$PSScriptRoot\TestHarness.psm1" -Force
+
+    # Extract the Write-Host override definition from the main script
+    # It's not a named function in the function definitions region — it's defined
+    # at script scope after the param block. We test it by dot-sourcing a minimal
+    # reproducer that matches the pattern in the main script.
+}
+
```
[the remark this answers is not in the document]
Comment (ZacharyLuz) [reply tests/WriteHost-Gating.Tests.ps1]: **Agree — fixed in commit `2718959`.**

The `Import-Module TestHarness` call and associated comments were removed. The test defines its Write-Host override inline (matching the main script''s pattern) and never needed TestHarness for extraction.

File: tests/ConvertFrom-Rest.Tests.ps1 [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+                restrictions = @(
+                    [PSCustomObject]@{
+                        type            = 'Zone'
+                        reasonCode      = 'NotAvailableForSubscription'
+                        restrictionInfo = [PSCustomObject]@{ zones = @('1', '2'); locations = @('eastus') }
+                    }
+                )
+                capabilities = @()
+            }
+            $result = ConvertFrom-RestSku -RestSku $restSku
```
[the remark this answers is not in the document]
Comment (ZacharyLuz) [reply tests/ConvertFrom-Rest.Tests.ps1]: **Disagree — false positive. Same Pester BeforeAll → It pattern as L29.**

`$result` is assigned in `BeforeAll` for the "SKU with restrictions" Context and used in the two `It` blocks that follow. PSScriptAnalyzer cannot cross scriptblock boundaries in Pester''s test structure.

File: tests/ConvertFrom-Rest.Tests.ps1 [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+        BeforeAll {
+            $restQuota = [PSCustomObject]@{
+                name = [PSCustomObject]@{
+                    value          = 'standardDSv3Family'
+                    localizedValue = 'Standard DSv3 Family vCPUs'
+                }
+                currentValue = 12
+                limit        = 100
+            }
+            $result = ConvertFrom-RestQuota -RestQuota $restQuota
```
[the remark this answers is not in the document]
Comment (ZacharyLuz) [reply tests/ConvertFrom-Rest.Tests.ps1]: **Disagree — false positive. Same Pester BeforeAll → It pattern as L29 and L82.**

`$result` in the "Standard quota normalization" Context is assigned in `BeforeAll` and consumed throughout the Context''s `It` blocks. PSScriptAnalyzer cannot trace this usage across scriptblock boundaries.

File: tests/WriteHost-Gating.Tests.ps1 [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+Describe "Write-Host Gating ($script:SuppressConsole)" {
+
+    BeforeAll {
+        # Define the override exactly as the main script does
+        $script:SuppressConsole = $false
+        function Write-Host {
```
[the remark this answers is not in the document]
Comment (ZacharyLuz) [reply tests/WriteHost-Gating.Tests.ps1]: **Intentional — this is the test subject.**

The test explicitly defines a `Write-Host` override to validate the gating behavior in isolation (without running the full main script). Defining the override in the test is the correct approach. The PSScriptAnalyzer rule fires correctly, but this is a deliberate override for testing purposes.

The `AfterAll { Remove-Item function:Write-Host }` cleanup added in commit `2718959` prevents session leak.

File: tests/WriteHost-Gating.Tests.ps1 [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            param(
+                [Parameter(Position = 0, ValueFromPipeline)]
+                [object]$Object = '',
```
[the remark this answers is not in the document]
Comment (ZacharyLuz) [reply tests/WriteHost-Gating.Tests.ps1]: **Partially agree — test-scope Write-Host proxy has same process-block gap. Accepted.**

The Write-Host override in the test mirrors the main script''s proxy exactly by design (the test is validating the proxy behavior). Adding a `process {}` block to the test override is unnecessary since `Write-Host` is never piped to in these tests. The AfterAll cleanup in `2718959` tears down the override after the test run.
